### DISCUSSION TITLE
Change tool installation directory

### DISCUSSION THREAD

**Discussion by Sn0wCrack** [shivammathur/setup-php#876] [Q&A / Help]
Right now all tools, such as composer, get installed globally to `/usr/local/bin`.

On self-hosted runner systems with multiple runner users, this can cause issues as it appears all tools are re-downloaded every time the action is executed.

Is there currently any way to change the install path to another directory, such as `~/.local/bin` or `/tmp`?

**Comment by Sn0wCrack** [comment]
Looking at the code I believe this may be hard coded currently: https://github.com/shivammathur/setup-php/blob/main/src/scripts/tools/add_tools.sh#L2

I can't seem to see anything else that overrides this value currently.

**Comment by shivammathur** [comment]
Can you reproduce the issue on a public repo and provide me the logs with `shivammathur/setup-php@verbose`. I would like to understand this better.

**Reply by Sn0wCrack** [reply]
I'll try and setup a minimal reproduction of this on a public repository later today or tomorrow.

To give you a rundown of our setup we have 4 VMs, each running 3 instances of the GitHub Runner software on this each:

 - Runner Server 1 - gitrunner01
   - Runner User 1 - runner01 - /home/runner01
   - Runner User 2 - runner02 - /home/runner02
   - Runner User 3 - runner03 - /home/runner03
 
- Runner Server 2 - gitrunner02
   - Runner User 1 - runner01 - /home/runner01
   - Runner User 2 - runner02 - /home/runner02
   - Runner User 3 - runner03 - /home/runner03

And so on...

With one application we actually build out a binary for various platforms using micro.sfx, which I do using a matrix for the target OS.

This is basically the only time in our standard workflows where multiple actions spawn at the exact same time on the same server, so we tend to only run into this issue here.

I've attached the logs of a run using the verbose version of setup-php you've mentioned here: [logs_55.zip](https://github.com/user-attachments/files/17499865/logs_55.zip)

And here's the workflow file we're currently using: [delpoyment.txt](https://github.com/user-attachments/files/17499880/delpoyment.txt)

**Reply by Sn0wCrack** [reply]
Apologies for the delay on this one, I've created a repository here that essentially replicates this issue:

https://github.com/Sn0wCrack/setup-php-multi-runner-users-reproduction

My runners are setup the exactly the same as mentioned above.

While I cannot replicate this as consistently as I can on the real runner servers we use compared to this one, it does trigger on occasion when the conditions are just right, such as these runs: 
- https://github.com/Sn0wCrack/setup-php-multi-runner-users-reproduction/actions/runs/11621624594/job/32365685642
- https://github.com/Sn0wCrack/setup-php-multi-runner-users-reproduction/actions/runs/11622040295/job/32366851229

My assumption has basically always been its just ultimately a race condition, if one job is trying to run composer and the other job is trying to download and install composer, it will error out like this.

@shivammathur

**Reply by shivammathur** [reply]
@Sn0wCrack 
I have tried to fix it here https://github.com/shivammathur/setup-php/commit/449afbcaeca797557bb91182b7303c090ed211b7
Please check if you can reproduce the error with `shivammathur/setup-php@develop`.

**Reply by Sn0wCrack** [reply]
@shivammathur 
Unfortunately this still occurs as the issue isn't with more than one job attempting to download the tool, but one downloading the tool while another job is using it.

**Reply by shivammathur** [reply]
@Sn0wCrack 
Please check again if https://github.com/shivammathur/setup-php/commit/80075aeba5b10e39a9b591da5dcd89fefaf61b89 fixes it.

**Reply by Sn0wCrack** [reply]
@shivammathur Unfortunately the issue still persists, however it does seem to happen far less frequently now.

**Reply by shivammathur** [reply]
@Sn0wCrack Please run it once again, I would like to see the logs.

**Reply by Sn0wCrack** [reply]
@shivammathur 

Here's a run where it failed: https://github.com/Sn0wCrack/setup-php-multi-runner-users-reproduction/actions/runs/11785460085

The last 3 runs should be of interest in this case.

**Reply by shivammathur** [reply]
It should install correctly now.

**Reply by Sn0wCrack** [reply]
@shivammathur That seems to be working now, last 5 runs seem to be good, thank you very much!

**Reply by Sn0wCrack** [reply]
Actually this might have been a fluke, it appears other processes may use composer at a later point and be kicked back with a "text file busy" error on composer.

https://github.com/Sn0wCrack/setup-php-multi-runner-users-reproduction/actions/runs/11788583187/job/32835929253

I still think it might be better to allow setting the tool directory (to say ~/.local/bin) so each runner user gets their own copy. Wasteful I know, but it will mean each one can act independently.
